tests/unit/experimental/test_scope_parser.py

```diff
@@ -0,0 +1,321 @@
+        # starting with open bracket
+        "[foo]",
+        # double brackets
+        "foo[[bar]]",
+        # unbalanced open brackets
+        "foo[",
+        "foo[bar",
+        # unbalanced close brackets
+        "foo]",
+        "foo bar]",
+        "foo[bar]]",
+        "foo[bar] baz]",
+        # space before brackets
+        "foo [bar]",
+        # missing space before next scope string after ']'
+        "foo[bar]baz",
+    ],
+)
+def test_scope_parsing_rejects_bad_inputs(scopestring):
+    with pytest.raises(ScopeParseError):
+        Scope.parse(scopestring)
+
+
+@pytest.mark.parametrize(
+    "scopestring",
+    [
+        "foo[foo]",
+        "foo[*foo]",
+        "foo[bar[foo]]",
+        "foo[bar[baz[bar]]]",
+        "foo[bar[*baz[bar]]]",
+        "foo[bar[*baz[*bar]]]",
+    ],
+)
+def test_scope_parsing_catches_and_rejects_cycles(scopestring):
+    with pytest.raises(ScopeCycleError):
+        Scope.parse(scopestring)
+
+
+def test_scope_parsing_catches_and_rejects_very_large_cycles_quickly():
+    """
+    WARNING: this test is hardware speed dependent and could fail on slow systems.
+
+    This test creates a very long cycle and validates that it can be caught in a
+    small timeframe of < 100ms.
+    Observed times on a test system were <20ms, and in CI were <60ms.
+
+    Although checking the speed in this way is not ideal, we want to avoid high
+    time-complexity in the cycle detection. This test offers good protection against any
+    major performance regression.
+    """
+    scope_string = ""
+    for i in range(1000):
+        scope_string += f"foo{i}["
+    scope_string += " foo10"
+    for _ in range(1000):
+        scope_string += "]"
+
+    t0 = time.time()
+    with pytest.raises(ScopeCycleError):
+        Scope.parse(scope_string)
+    t1 = time.time()
+    assert t1 - t0 < 0.1
+
+
+@pytest.mark.parametrize(
+    "scopestring",
+    ("foo", "*foo", "foo[bar]", "foo[*bar]", "foo bar", "foo[bar[baz]]"),
+)
+def test_scope_parsing_accepts_valid_inputs(scopestring):
+    # test *only* that parsing does not error and returns a non-empty list of scopes
+    scopes = Scope.parse(scopestring)
+    assert isinstance(scopes, list)
+    assert len(scopes) > 0
+    assert isinstance(scopes[0], Scope)
+
+
+def test_scope_deserialize_simple():
+    scope = Scope.deserialize("foo")
+    assert str(scope) == "foo"
+
+
+def test_scope_deserialize_with_dependencies():
+    # oh, while we're here, let's also check that our whitespace insensitivity works
+    scope = Scope.deserialize("foo[ bar   *baz  ]")
+    assert str(scope) in ("foo[bar *baz]", "foo[*baz bar]")
+
+
+def test_scope_deserialize_fails_on_empty():
+    with pytest.raises(ValueError):
+        Scope.deserialize("  ")
+
+
+def test_scope_deserialize_fails_on_multiple_top_level_scopes():
+    with pytest.raises(ValueError):
+        Scope.deserialize("foo bar")
+
+
+@pytest.mark.parametrize("scope_str", ("*foo", "foo[bar]", "foo[", "foo]", "foo bar"))
```

Comment:
`"foo bar"` doesn't appear to relate to special character tests, and appears to be tested immediately above...although now I need to go back and review the difference between `Scope.deserialize(s)` and `Scope(s)`.

Reply:
The behaviors tested are slightly different. Here are the three relevant methods at a high level:

1. `Scope.parse` does a full parse and provides a list of `Scope`s as the result.
2. `Scope.deserialize` runs `Scope.parse` and then asserts that the length of the resulting list is exactly 1 -- and it unwraps that value.
3. `Scope.__init__` does not do any parsing. It takes any input string and assigns it to the `Scope.scope_string` attribute, after checking that it was a valid value to use

`__init__` does no parsing and rejects these strings because this is a simpler design than some alternatives for breaking the potential for a `parse -> __init__ -> parse` cycle. (I considered the path of a true alternate constructor which does not call `__init__`, but I thought that wasn't worth it in the short term.)

So this test ensures that `Scope("foo bar")` is rejected "because you can't initialize a scope whose string contains a space anywhere".

By contrast, the above test is checking that "if you deserialize a scope string, it must parse such that there is only one top-level or 'root' scope". Spaces are allowed in cases like `Scope.deserialize("foo[bar baz]")`.